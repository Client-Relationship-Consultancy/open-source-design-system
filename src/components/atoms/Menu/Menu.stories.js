/* eslint no-param-reassign: 0 */

import React from "react"

import { storiesOf } from "@storybook/react"

import { Menu } from "./Menu"

const stories = storiesOf("Menu", module)

stories.add("Menu", () => {
  return (
    <div style={{ paddingBottom: "5rem" }}>
      <Menu />
    </div>
  )
})
