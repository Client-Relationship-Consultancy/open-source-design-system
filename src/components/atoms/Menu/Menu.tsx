import React from "react";
import styled from "styled-components";

import { colourPalette } from "../../../brandColours";

const defaultPalette = colourPalette.examplePalette;

export class Menu extends React.Component {
  render = () => {
    return <div>hello</div>;
  };
}
