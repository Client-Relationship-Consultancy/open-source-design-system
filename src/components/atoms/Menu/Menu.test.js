import React from "react"
import { shallow, mount } from "enzyme"
import toJson from "enzyme-to-json"
import { Menu, StyledMenu, MenuLabel, SubMenu, SubMenuItems } from "./Menu"

describe("Menu Component", () => {
  const commonProps = {
    items: [
      { label: "Exclude me", id: "exclude", onClick: jest.fn(), icon: "users" },
      { label: "Remove me", onClick: jest.fn(), icon: <svg /> },
      { label: "Activate me", id: "activate", onClick: jest.fn() },
    ],
  }

  let component
  beforeEach(() => {
    document.addEventListener = jest.fn()
    document.removeEventListener = jest.fn()

    component = shallow(<Menu {...commonProps}>Hello</Menu>)
  })

  it("should match snapshot", () => {
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it("should render the arrow icon if showMenuArrow is true", () => {
    const dropdownIcon = component.find("Icon[name='chevron-down']")
    expect(dropdownIcon.exists()).toEqual(true)
    expect(dropdownIcon.length).toEqual(1)
  })

  it("should not render the arrow icon if showMenuArrow is false", () => {
    component.setProps({ showMenuArrow: false })
    const dropdownIcon = component.find("Icon[name='chevron-down']")
    expect(dropdownIcon.exists()).toEqual(false)
  })

  it("should toggle the showSubMenu state when the button is clicked", () => {
    expect(component.state()).toEqual({
      showSubMenu: false, // false by default
    })

    component.simulate("click")
    expect(component.state()).toEqual({
      showSubMenu: true,
    })

    component.simulate("click")
    expect(component.state()).toEqual({
      showSubMenu: false,
    })
  })

  it("should toggle the showSubMenu state when openSubMenu is called", () => {
    expect(component.state()).toEqual({
      showSubMenu: false, // false by default
    })

    component.instance().openSubMenu()
    expect(component.state()).toEqual({
      showSubMenu: true,
    })

    component.instance().openSubMenu()
    expect(component.state()).toEqual({
      showSubMenu: false,
    })
  })

  it("should close the menu when the current component ref returns false as it does not contain the event target", () => {
    component.setState({ showSubMenu: true })
    expect(component.state()).toEqual({ showSubMenu: true })

    const mockContains = jest.fn(() => false)
    component.instance().componentRef = {
      current: { contains: mockContains },
    }

    const mockMouseEvent = {
      target: null,
    }
    component.instance().closeMenuWhenClickedOutside(mockMouseEvent)
    expect(mockContains).toHaveBeenCalledWith(null)
    expect(component.state()).toEqual({ showSubMenu: false })
  })

  it("should leave the menu open when the current component ref returns true as it does contain the event target", () => {
    component.setState({ showSubMenu: true })
    expect(component.state()).toEqual({ showSubMenu: true })

    const mockContains = jest.fn(() => true)
    component.instance().componentRef = {
      current: { contains: mockContains },
    }

    const mockMouseEvent = {
      target: <div />,
    }
    component.instance().closeMenuWhenClickedOutside(mockMouseEvent)
    expect(mockContains).toHaveBeenCalledWith(<div />)
    expect(component.state()).toEqual({ showSubMenu: true })
  })

  it("should add the event listener when the component mounts", () => {
    expect(document.addEventListener).toHaveBeenCalledWith(
      "mousedown",
      component.instance().closeMenuWhenClickedOutside,
    )
  })

  it("should remove the event listener when the component unmounts", () => {
    const eventListener = component.instance().closeMenuWhenClickedOutside
    component.unmount()
    expect(document.removeEventListener).toHaveBeenCalledWith("mousedown", eventListener)
  })

  it("should use the ref prop when useInnerRef prop is false", () => {
    component.setProps({
      useInnerRef: false,
    })
    expect(component.instance().createRefProp()).toEqual({
      ref: component.instance().componentRef,
    })
  })

  it("should use the innerRef prop when useInnerRef prop is true", () => {
    component.setProps({
      useInnerRef: true,
    })
    expect(component.instance().createRefProp()).toEqual({
      innerRef: component.instance().componentRef,
    })
  })
})

describe("StyledMenu component", () => {
  let component
  beforeEach(() => {
    const mockTheme = { action: { main: { hex: "mockHexValue" } } }
    component = mount(<StyledMenu showSubMenu theme={mockTheme} />)
  })

  it("should match snapshot", () => {
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})

describe("MenuLabel component", () => {
  it("should match snapshot", () => {
    const component = mount(<MenuLabel />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})

describe("SubMenu component", () => {
  let component
  beforeEach(() => {
    const mockTheme = {
      surface: { hex: "mockSurfaceHex" },
      black: { tint40: { hex: "mockBlackTint40" } },
    }
    component = mount(<SubMenu showSubMenu theme={mockTheme} />)
  })

  it("should match snapshot", () => {
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it("should set opacity to be 1 and visibility to visible when showSubMenu is true", () => {
    expect(component).toHaveStyleRule("visibility", "visible")
    expect(component).toHaveStyleRule("opacity", "1")
  })

  it("should set opacity to be 1 and visibility to hidden when showSubMenu is true", () => {
    component.setProps({ showSubMenu: false })
    expect(component).toHaveStyleRule("opacity", "0")
    expect(component).toHaveStyleRule("visibility", "hidden")
  })
})

describe("SubMenuItems component", () => {
  const mockTheme = {
    black: { tint80: { hex: "mockBlackTint80" }, tint20: { hex: "mockBlackTint20" } },
  }
  it("should match snapshot", () => {
    const component = mount(<SubMenuItems showSubMenu theme={mockTheme} />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})
