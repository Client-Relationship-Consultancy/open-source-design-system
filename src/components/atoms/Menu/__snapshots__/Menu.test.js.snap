// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Component should match snapshot 1`] = `
<Menu__StyledMenu
  onClick={[Function]}
  showSubMenu={false}
>
  <Menu__MenuLabel>
    <span>
      Hello
    </span>
    <Icon
      color="action"
      name="chevron-down"
    />
  </Menu__MenuLabel>
  <Menu__SubMenu
    className="subMenu"
    showSubMenu={false}
  >
    <Menu__SubMenuItems
      key="exclude"
      onClick={[MockFunction]}
    >
      <Icon
        name="users"
      />
      <span>
        Exclude me
      </span>
    </Menu__SubMenuItems>
    <Menu__SubMenuItems
      key="Remove me"
      onClick={[MockFunction]}
    >
      <svg />
      <span>
        Remove me
      </span>
    </Menu__SubMenuItems>
    <Menu__SubMenuItems
      key="activate"
      onClick={[MockFunction]}
    >
      <span>
        Activate me
      </span>
    </Menu__SubMenuItems>
  </Menu__SubMenu>
</Menu__StyledMenu>
`;

exports[`MenuLabel component should match snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0.5rem;
}

.c0 > * + * {
  margin-left: 0.75rem;
}

.c0:hover {
  cursor: pointer;
}

<Menu__MenuLabel>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Menu__MenuLabel-wb6ftp-1",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "display:flex;padding:0.5rem;> * + *{margin-left:0.75rem;}:hover{cursor:pointer;}",
          ],
        },
        "displayName": "Menu__MenuLabel",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Menu__MenuLabel-wb6ftp-1",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <div
      className="c0"
    />
  </StyledComponent>
</Menu__MenuLabel>
`;

exports[`StyledMenu component should match snapshot 1`] = `
.c0 {
  color: mockHexValue;
  font-family: "Gentona","Montserrat";
  font-size: 0.875rem;
  position: relative;
  background-color: transparent;
  border: none;
  outline: none;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  text-align: center;
}

<Menu__StyledMenu
  showSubMenu={true}
  theme={
    Object {
      "action": Object {
        "main": Object {
          "hex": "mockHexValue",
        },
      },
    }
  }
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Menu__StyledMenu-wb6ftp-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "color:",
            [Function],
            ";font-family:\\"Gentona\\",\\"Montserrat\\";font-size:0.875rem;position:relative;background-color:transparent;border:none;outline:none;transition:all 0.3s ease;text-align:center;",
          ],
        },
        "displayName": "Menu__StyledMenu",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Menu__StyledMenu-wb6ftp-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    showSubMenu={true}
    theme={
      Object {
        "action": Object {
          "main": Object {
            "hex": "mockHexValue",
          },
        },
      }
    }
  >
    <div
      className="c0"
    />
  </StyledComponent>
</Menu__StyledMenu>
`;

exports[`SubMenu component should match snapshot 1`] = `
.c0 {
  min-width: 110%;
  margin-top: 0.1rem;
  position: absolute;
  max-width: 500%;
  white-space: nowrap;
  background-color: mockSurfaceHex;
  box-shadow: 0.1rem 0.1rem 0.5rem mockBlackTint40;
  visibility: visible;
  opacity: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: all 0.3s ease;
  transition: all 0.3s ease;
  border-radius: 0.2rem;
  padding: 0.25rem 0;
  z-index: 10;
}

<Menu__SubMenu
  showSubMenu={true}
  theme={
    Object {
      "black": Object {
        "tint40": Object {
          "hex": "mockBlackTint40",
        },
      },
      "surface": Object {
        "hex": "mockSurfaceHex",
      },
    }
  }
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Menu__SubMenu-wb6ftp-2",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "min-width:110%;margin-top:0.1rem;position:absolute;max-width:500%;white-space:nowrap;background-color:",
            [Function],
            ";box-shadow:0.1rem 0.1rem 0.5rem ",
            [Function],
            ";visibility:",
            [Function],
            ";opacity:",
            [Function],
            ";display:flex;flex-direction:column;transition:all 0.3s ease;border-radius:0.2rem;padding:0.25rem 0;z-index:10;",
          ],
        },
        "displayName": "Menu__SubMenu",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Menu__SubMenu-wb6ftp-2",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    showSubMenu={true}
    theme={
      Object {
        "black": Object {
          "tint40": Object {
            "hex": "mockBlackTint40",
          },
        },
        "surface": Object {
          "hex": "mockSurfaceHex",
        },
      }
    }
  >
    <div
      className="c0"
    />
  </StyledComponent>
</Menu__SubMenu>
`;

exports[`SubMenuItems component should match snapshot 1`] = `
.c0 {
  background-color: transparent;
  outline: none;
  border: none;
  padding: 0.5rem 0.75rem;
  width: 100%;
  color: mockBlackTint80;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c0:hover {
  cursor: pointer;
  background-color: mockBlackTint20;
}

.c0 > * + * {
  margin-left: 0.5rem;
}

.c0 svg {
  max-width: 1rem;
}

<Menu__SubMenuItems
  showSubMenu={true}
  theme={
    Object {
      "black": Object {
        "tint20": Object {
          "hex": "mockBlackTint20",
        },
        "tint80": Object {
          "hex": "mockBlackTint80",
        },
      },
    }
  }
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Menu__SubMenuItems-wb6ftp-3",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "background-color:transparent;outline:none;border:none;padding:0.5rem 0.75rem;width:100%;color:",
            [Function],
            ";display:flex;text-align:left;transition:all 0.3s linear;:hover{cursor:pointer;background-color:",
            [Function],
            ";}> * + *{margin-left:0.5rem;}svg{max-width:1rem;}",
          ],
        },
        "displayName": "Menu__SubMenuItems",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Menu__SubMenuItems-wb6ftp-3",
        "target": "button",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    showSubMenu={true}
    theme={
      Object {
        "black": Object {
          "tint20": Object {
            "hex": "mockBlackTint20",
          },
          "tint80": Object {
            "hex": "mockBlackTint80",
          },
        },
      }
    }
  >
    <button
      className="c0"
    />
  </StyledComponent>
</Menu__SubMenuItems>
`;
